### PR TITLE
algorithms: Add bhalf rand support

### PR DESCRIPTION
**Pull request by e10harvey**
Supersedes #4652.
Related to #4372.

### CODE REVIEW

**Comment by e10harvey** [comment]
retest this please

File: algorithms/unit_tests/TestRandom.hpp [hunk 1]
```diff
@@ -424,6 +432,12 @@ struct test_random_scalar {
       }
 #endif
 
+#if defined(KOKKOS_BHALF_T_IS_FLOAT) && !KOKKOS_BHALF_T_IS_FLOAT
+      if (std::is_same<Scalar, Kokkos::Experimental::bhalf_t>::value) {
+        variance_factor = 15.01;
```
Comment (dalg24) [inline algorithms/unit_tests/TestRandom.hpp:437]: Where does this number come from?

File: algorithms/unit_tests/TestRandom.hpp [hunk 2]
```diff
@@ -386,6 +386,14 @@ struct test_random_scalar {
       }
 #endif
 
+#if defined(KOKKOS_BHALF_T_IS_FLOAT) && !KOKKOS_BHALF_T_IS_FLOAT
+      if (std::is_same<Scalar, Kokkos::Experimental::bhalf_t>::value) {
+        mean_eps_expect       = 0.019;
+        variance_eps_expect   = 1.0;
+        covariance_eps_expect = 2.8e4;
```
Comment (dalg24) [inline algorithms/unit_tests/TestRandom.hpp:393]: How did you determine these numbers?

File: algorithms/unit_tests/TestRandom.hpp (hunk 1 above)
Comment (e10harvey) [reply algorithms/unit_tests/TestRandom.hpp:437]: Based on the test failure output with a smaller variance factor, this is approximately the smallest variance factor I found that allows the tests to pass.

File: algorithms/unit_tests/TestRandom.hpp (hunk 2 above)
Comment (e10harvey) [reply algorithms/unit_tests/TestRandom.hpp:393]: Same as above.

**Comment by e10harvey** [comment]
@crtrott Is there a way to calculate the expected variance factor and epsilons based on the scalar precision?

**Comment by e10harvey** [comment]
retest this please

**Comment by ajpowelsnl** [comment]
Retest this please

**Comment by crtrott** [comment]
There is certainly a way to calculate where these numbers come from, but its likely pretty complicated because the semi randomness of roundof errors plays a role here, which for half and bhalf are not enough orders of magnitude away from the actual random generator randomness. Unfortunately I doubt its a fully uncorrelated distribution which makes this all iffy. I don't have time to work the math out at least.

**Comment by dalg24** [comment]
Retest this please

**Review by dalg24 — APPROVED** [review]

**Comment by rgayatri23** [comment]
The OpenMPTarget failure is unrelated to this issue. Seems like the OpenMPTarget CI is having issues.

**Comment by e10harvey** [comment]
@crtrott: Ready to merge?

**Review by crtrott — APPROVED** [review]

**Pull request merged**